Routing/multiple table ranges

aaaaaaaalex opened the pull request:
## Description

<!-- A few sentences describing the overall goals of the pull request's commits.
Please include
- the type of fix - (e.g. bug fix, new feature, documentation)
- some details on _why_ this PR should be merged
- the details of the testing you've done on it (both manual and automated)
- which components are affected by this PR
- links to issues that this PR addresses
-->

Adds support for multiple RT table-ranges
[Modern Linux kernels support >255 routing tables](https://github.com/torvalds/linux/blob/master/include/uapi/linux/rtnetlink.h#L338), this change aims to make them available to Felix

## Related issues/PRs

<!-- If appropriate, include a link to the issue this fixes.
fixes <ISSUE LINK>

If appropriate, add links to any number of PRs documented by this PR
documents <PR LINK>
-->

## Todos

- [x] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

<!-- Writing a release note:
- By default, no release note action is required.
- If you're unsure whether or not your PR needs a note, ask your reviewer for guidance.
- If this PR requires a release note, update the block below to include a concise note describing
  the change and any important impacts this PR may have.
-->

```release-note
Deprecated config arg `routeTableRange` in favour of `routeTableRanges`. Felix can now make use of a larger set of Linux route-tables, and can be configured to avoid sensitive ranges. This implicitly drops support for very old Linux kernels (below v2.6)
```

mikestephen commented on calico/reference/felix/configuration.md:
```diff
@@ -69,7 +69,7 @@ The full list of parameters which can be set is as follows.
 | `ReportingIntervalSecs`           | `FELIX_REPORTINGINTERVALSECS`           | Interval at which Felix reports its status into the datastore or `0` to disable. Must be non-zero in OpenStack deployments. [Default: `30`] | int |
 | `ReportingTTLSecs`                | `FELIX_REPORTINGTTLSECS`                | Time-to-live setting for process-wide status reports. [Default: `90`] | int |
-| `RouteTableRange`                 | `FELIX_ROUTETABLERANGE`                 | Calico programs additional Linux route tables for various purposes.  `RouteTableRange` specifies the indices of the route tables that Calico should use. [Default: `1-250`] | `<min>-<max>` |
+| `RouteTableRanges`                 | `FELIX_ROUTETABLERANGES`                 | Calico programs additional Linux route tables for various purposes. `RouteTableRanges` designates a set of table ranges that Calico is permitted to use. [Default: `1-250`] | `<min>-<max>,<min>-<max>,...` |
```
Should the range here match the below value of `1-250,256-10000`?

aaaaaaaalex replied:
yes it should, good catch!

aaaaaaaalex replied:
or alternatively, I could change the other one to be 1-250, to maintain backward compat :thinking:

mikestephen replied:
Yeah, you're right, probably something we need to get @fasaxc 's perspective on.

aaaaaaaalex replied:
final approach is to specify `1-10000` and let felix ignore reserved tables

mikestephen commented on calico/reference/resources/felixconfig.md:
```diff
@@ -131,12 +131,19 @@ policy is always accelerated, using the best available BPF technology.
+| min      | Minimum index to use | int    |
+| max      | Maximum index to use | int    |
+
+Each `RouteTableRange` designates a set of routing tables available to Calico. Modern Linux kernels can theorectically support more than the documented 252 additional tables. However, allocating more than the default number of tables is not as simple as increasing the max table index, since the increased range would encpomass the reserved tables, and could lead to Calico clashing with the kernel.
```
Typo in `encpomass`.

aaaaaaaalex replied:
nice catch!

mikestephen commented on felix/config/param_types.go:
```diff
@@ -554,28 +554,62 @@ func (r *RegionParam) Parse(raw string) (result interface{}, err error) {
 
-type RouteTableRangeParam struct {
+// reserved linux kernel routing tables (cannot be targeted by routeTableRanges)
+var routeTablesReservedLinux = []int{253, 254, 255}
```
According to [this](http://linux-ip.net/html/routing-tables.html) it might also be wise to reserve table 0, though I'm not sure if other logic already excludes 0.

aaaaaaaalex replied:
0 is the empty (unspecified) val I believe? I assumed that value gets parsed as "none" , but will dig into it and confirm :slightly_smiling_face:

fasaxc commented on calico/reference/resources/felixconfig.md:
```diff
@@ -132,12 +132,19 @@ policy is always accelerated, using the best available BPF technology.
+| min      | Minimum index to use | int    |
+| max      | Maximum index to use | int    |
+
+Each `RouteTableRange` designates a set of routing tables available to Calico. Modern Linux kernels can theorectically support more than the documented 252 additional tables. However, allocating more than the default number of tables is not as simple as increasing the max table index, since the increased range would encompass the reserved tables, and could lead to Calico clashing with the kernel.
```
Can live with this approach (we've got lots of config parameters that are similarly handled), but (with experience gained from all those parameters...) I think it'd be better for the user if you just automatically excluded the reserved Linux tables.  I.e. let the user specify `10-10000` and just discard the `25x` tables with an INFO log `"Route table range contains linux reserved range, excluding 253-255."`.  

Then I'd include a note in the docs that Felix automatically excludes `0 and 253-255` since they are reserved by Linux to let knowledgeable users know so they don't freak out.   Rule of thumb: if the explanation in the docs can be shorter/easier, it might be a good idea.

aaaaaaaalex replied:
That does sound like a favourable solution in fairness! I think the main reason I opted for this approach is that different provisioners will have different reserved tables - E.g, in the case of AWS, as noted in the docs. I'm not hugely familiar with how much variance there is across provisioners, but maintaining a list of which ones reserve which tables gave me a hunch it may be better to go with the verbose option.
Totally makes sense what you're saying though, and my worries may be overblown! :slightly_smiling_face:

fasaxc replied:
I was only thinking of ignoring the Linux-reserved tables.  Will still need config for other provisioners.

fasaxc commented on felix/config/param_types.go:
```diff
@@ -554,28 +554,63 @@ func (r *RegionParam) Parse(raw string) (result interface{}, err error) {
+
+		// check if ranges collide with reserved linux tables
+		for _, rsrv := range routeTablesReservedLinux {
+			if min == rsrv {
```
Why not just 
```
if min <= rsrv && max >= rsrv {
  ...
}
```

fasaxc commented on libcalico-go/lib/validator/v3/validator.go:
```diff
@@ -1560,18 +1564,54 @@ func validateRuleMetadata(structLevel validator.StructLevel) {
+
+	// check if ranges collide with reserved linux tables
+	for _, rsrv := range routeTablesReservedLinux {
+		if r.Min == rsrv || (r.Min < rsrv && r.Max >= rsrv) {
```
Why not use the `<=` operator?

aaaaaaaalex replied:
Had a flu all last week, going to blame that 😆

fasaxc reviewed (changes requested):
Generally looking good, with good tests etc.  One major issue that needs to be addressed  (changing the config param will break back compatibility) and a few more minor thoughts.

aaaaaaaalex commented on felix/idalloc/index_allocator.go:
```diff
@@ -26,24 +27,57 @@ type IndexRange struct {
-		r.indexStack.Push(i)
+
+	var lowestIndex int
+	for j, indexRange := range indexRanges {
```
should we cap it at, say, 100,000?

aaaaaaaalex replied:
changed to `0xFFFF` for now, ~65000

aaaaaaaalex commented on felix/config/param_types.go:
```diff
@@ -554,28 +554,63 @@ func (r *RegionParam) Parse(raw string) (result interface{}, err error) {
+		}
+
+		if int64(max) > int64(routeTableMaxLinux) {
+			err = p.parseFailed(raw, "max value is too high")
```
i was thinking based on the regex above that negative numbers would not be a possible input, can throw it in anyway though just incase the regex changes down the line :ok_hand:

aaaaaaaalex commented on api/pkg/apis/projectcalico/v3/felixconfig.go:
```diff
@@ -366,9 +366,9 @@ type FelixConfigurationSpec struct {
+	// Calico programs additional Linux route tables for various purposes.  RouteTableRanges
 	// specifies the indices of the route tables that Calico should use.
-	RouteTableRange *RouteTableRange `json:"routeTableRange,omitempty" validate:"omitempty"`
+	RouteTableRanges *RouteTableRanges `json:"routeTableRanges,omitempty" validate:"omitempty,dive"`
```
If I were to preserve the old name, and make sure any previous-version value worked in the new arg as normal, would that work?

fasaxc replied:
I think you'd run into problems in the API structs, it'll break the Golang API even if you did the work to still accept the exact same YAML?

fasaxc replied:
(That'd have knock-on effects for users and for the operator I think).

aaaaaaaalex replied:
ahh ok, didn't consider that, will add back in the old param too, thanks :+1:

fasaxc commented on libcalico-go/lib/validator/v3/validator.go:
```diff
@@ -45,6 +45,9 @@ const (
 	// Maximum size of annotations.
 	totalAnnotationSizeLimitB int64 = 256 * (1 << 10) // 256 kB
 
+	// linux can support route-tables with indices up to 0xfffffff, however, using all of them would likely blow up, so cap the limit at 65535
```
How about capping the size of a single _range_ but allowing that range to be placed anywhere in the 32-bits?  That seems like a happy medium.

aaaaaaaalex replied:
nice, sounds good

aaaaaaaalex commented on felix/idalloc/index_allocator.go:
```diff
@@ -26,24 +27,72 @@ type IndexRange struct {
+			// skip exclusions
+			for _, excl := range exclusions {
+				if excl.contains(i) {
+					continue populating
```
now ignoring routetables here, based on `exclusions` passed to the indexAllocator - I also log.Info when the same is noticed in validator and param_types.go 
CC @fasaxc

fasaxc commented on calico/reference/felix/configuration.md:
```diff
@@ -69,7 +69,8 @@ The full list of parameters which can be set is as follows.
 | `ReportingIntervalSecs`           | `FELIX_REPORTINGINTERVALSECS`           | Interval at which Felix reports its status into the datastore or `0` to disable. Must be non-zero in OpenStack deployments. [Default: `30`] | int |
 | `ReportingTTLSecs`                | `FELIX_REPORTINGTTLSECS`                | Time-to-live setting for process-wide status reports. [Default: `90`] | int |
-| `RouteTableRanges`                 | `FELIX_ROUTETABLERANGES`                 | Calico programs additional Linux route tables for various purposes. `RouteTableRanges` designates a set of table ranges that Calico is permitted to use. [Default: `1-250,255-10000`] | `<min>-<max>,<min>-<max>,...` |
+| `RouteTableRange`                 | `FELIX_ROUTETABLERANGE`                 | *deprecated in favour of `RouteTableRanges`* Calico programs additional Linux route tables for various purposes.  `RouteTableRange` specifies the indices of the route tables that Calico should use. [Default: `1-250`] | `<min>-<max>` |
```
Is "Default 1-250" right now?  I'd expect the default to be unset so that `RouteTableRanges` takes precidence

fasaxc commented on calico/reference/resources/felixconfig.md:
```diff
@@ -82,7 +82,8 @@ spec:
 | reportingTTL                       | Time-to-live setting for process-wide status reports. | `5s`, `10s`, `1m` etc. | duration | `90s` |
 | routeRefreshInterval               | Period at which Felix re-checks the routes in the dataplane to ensure that no other process has accidentally broken {{site.prodname}}'s rules. Set to 0 to disable route refresh. | `5s`, `10s`, `1m` etc. | duration | `90s` |
-| routeTableRanges                    | Calico programs additional Linux route tables for various purposes. `RouteTableRanges` designates a set of table ranges that Calico is permitted to use. |  | [RouteTableRanges](#routetableranges) | `[{Min: 1, Max: 250}, {Min: 256, Max: 10000}]` |
+| routeTableRange                    | *deprecated in favour of `RouteTableRanges`* Calico programs additional Linux route tables for various purposes.  `RouteTableRange` specifies the indices of the route tables that Calico should use. |  | [RouteTableRanges](#routetablerange) | `{Min: 1, Max: 250}` |
```
Same here, should the default be unset?

fasaxc commented on calico/reference/felix/configuration.md:
```diff
@@ -69,7 +69,8 @@ The full list of parameters which can be set is as follows.
 | `ReportingTTLSecs`                | `FELIX_REPORTINGTTLSECS`                | Time-to-live setting for process-wide status reports. [Default: `90`] | int |
-| `RouteTableRanges`                 | `FELIX_ROUTETABLERANGES`                 | Calico programs additional Linux route tables for various purposes. `RouteTableRanges` designates a set of table ranges that Calico is permitted to use. [Default: `1-250,255-10000`] | `<min>-<max>,<min>-<max>,...` |
+| `RouteTableRange`                 | `FELIX_ROUTETABLERANGE`                 | *deprecated in favour of `RouteTableRanges`* Calico programs additional Linux route tables for various purposes.  `RouteTableRange` specifies the indices of the route tables that Calico should use. [Default: `1-250`] | `<min>-<max>` |
+| `RouteTableRanges`                 | `FELIX_ROUTETABLERANGES`                 | Calico programs additional Linux route tables for various purposes. `RouteTableRanges` designates a set of table ranges that Calico is permitted to use. Overrides `RouteTableRange` if both are present. [Default: `1-250,255-10000`] | `<min>-<max>,<min>-<max>,...` |
```
Looks like you've changed the actual default to 1-10000 but this doc still says the two groups

fasaxc commented on felix/config/param_types.go:
```diff
@@ -558,21 +558,48 @@ func (r *RegionParam) Parse(raw string) (result interface{}, err error) {
-	if m == nil {
+	match := regexp.MustCompile(`(\d+)-(\d+)`).FindAllStringSubmatch(raw, -1)
+	if match == nil {
 		err = p.parseFailed(raw, "must be a list of route-table ranges which do not designate reserved tables")
```
Think you've removed the "reserved tables" restriction now

fasaxc replied:
So, shouldn't the error be changed?

fasaxc commented on felix/config/config_params.go:
```diff
@@ -331,7 +331,9 @@ type Config struct {
 
-	RouteTableRange idalloc.IndexRange `config:"route-table-range;1-250;die-on-fail"`
+	// RouteTableRange is deprecated in favour of RouteTableRanges
+	RouteTableRange  idalloc.IndexRange   `config:"route-table-range;1-250;die-on-fail"`
```
Feel like one of these should have no default so that:
* if the user has already specified `RouteTableRange` in their config, we honour that.
* Otherwise, we use the new `RouteTableRanges` value

aaaaaaaalex replied:
In the case where both are set, presumably `RouteTableRanges` should take precedence, however, is there a way of differentiating between a default-set value and a user-set value?

The reason I ask is because if I remove the old arg's default in favour of an empty val...
 - I can always see when the old arg is explicitly-set, but
 - can only tell when the new arg is non-default.
 - if for some reason the user set the new arg explicitly to the same value as default, but also has the old arg set to something, it would be ambiguous which value I should run with. I _should_ be taking the new one, but it _looks_ like a default val...

Hope this is making sense... Im wondering should the _old_ arg be the one to take precedence, but also have no default. This would disambiguate the situation, but Im not sure if It's the right call.

caseydavenport replied:
We should probably start moving to a world where defaults are done in the API and not in felix code.

Doing it in code means upgrades can break users more easily (because an implicit default changed). Doing it via the API means changing felix out won't change the behavior of the cluster.

I guess sometimes we want to be able to sneakily change something, but in general that's bad practice and imo would be better to be explicit about the config running on the system.

This is how we handle defaulting for all of the operator APIs, for example

fasaxc commented on felix/config/config_params.go:
```diff
@@ -331,7 +331,9 @@ type Config struct {
-	RouteTableRange idalloc.IndexRange `config:"route-table-range;1-250;die-on-fail"`
+	// RouteTableRange is deprecated in favour of RouteTableRanges
+	RouteTableRange  idalloc.IndexRange   `config:"route-table-range;1-250;die-on-fail"`
+	RouteTableRanges []idalloc.IndexRange `config:"route-table-ranges;1-10000;die-on-fail"`
```
I think you might have missed the actual defaulting logic for the case where `RouteTableRanges` is not set.  I'd add a method to the ConfigParams object that checks which of the two fields is set explicitly and returns an `[]idalloc.IndexRange` in both cases.

fasaxc reviewed (changes requested):
Thanks Alex,

aaaaaaaalex commented:
Think Im getting tunnel vision, yes, totally missed the actual defaulting logic :man_facepalming: will fix up the default vals. Also, yes I've removed the reserved tables functionality - instead, I modified the `index_allocator` to accept a list of `exclusions`, where I define the reserved indices, which it will then skip when creating its index stack

aaaaaaaalex commented on calico/_includes/content/install-awscni-routetable-issue.md:
````diff
@@ -1,13 +1,13 @@
 > 1. Configure Felix to manage a routing table range which is distinct from the range used by AWS CNI:
 >     ```bash
->     kubectl patch felixconfiguration default --type='merge' -p '{"spec": {"routeTableRange":{"min": 65, "max": 99}}}'
+>     kubectl patch felixconfiguration default --type='merge' -p '{"spec": {"routeTableRanges":[{"min": 65, "max": 99}, {"min": 256, "max": 10000}]}}'
````
need to test this

fasaxc commented on calico/reference/resources/felixconfig.md:
```diff
@@ -150,12 +150,16 @@ The `RouteTanbleRange` option is now deprecated in favour of [RouteTableRanges](
+*Note*, for performance reasons, the maximum total number of routing tables that Felix will accept is 65535 (or 2*16).
 
-If `RouteTableRange` and `RouteTableRanges` are both present, the latter will take precedence.
+If the deprecated `RouteTableRange` option is explicitly set, it will be honoured in favour of `RouteTableRanges`.
```
I think I'd have jumped the other way if both are set.  If both are explicitly set, I'd use the value from RouteTableRanges.

aaaaaaaalex replied:
I initially wanted to go that route, but I encountered a problem:
 - remove `RouteTableRange` default so we can check when it's empty
 - `RouteTableRanges` still has a non-zero default, so it will always look "set"-ish
 - if the user the deprecated arg and left the new one as default, while I _could_ check that `RouteTableRanges` arg isnt _default_, and go with the deprecated arg, there would be ambiguity - if they for some reason explicitly set `RouteTableRanges` to the default val, I wouldn't be able to tell that apart :thinking: 
 
I didn't immediately see a way from within the config function to differentiate between set (to the default) and unset default, so I opted to give the old param precedence when it's there, since if it's unset, it'll definitely be a zero-val

Hope that makes sense. maybe Im missing a way around this?

fasaxc replied:
One option would be to remove both their defaults in the Felix config_params.go and then do the defaulting "manually" in the getter function.  We do it that way to handle a couple of other "more complicated" defaults.

aaaaaaaalex replied:
Thats a great suggestion :raised_hands: will implement.

caseydavenport replied:
I believe the way that Kubernetes has handled API changes from singular to plural is through validation that both are set consistently.

e.g., if `RouteTableRange` and `RouteTableRanges` are set, they need to be consistent with each other else they will error. 

That might just mean "Return an error on the API if both are set" - that seems like the cleanest behavior to me.

caseydavenport replied:
The way I think about this is that the code shouldn't be assuming or trying to guess the users intent - if the user gives us two pieces of conflicting intent via the API, we need to tell the user "Hey, you're not making sense".

aaaaaaaalex replied:
I like the suggestion, personally am happy with either approach. Are there other examples of either solution being done before in our code? If this is the first occurence, I'm happy to put the best foot forward, but Im also conscious of getting choice paralysis here as I'd be rewriting this piece of the PR for the third time :laughing:  if it's something that we're ambiguous about, I might acknowledge the suggestion but leave it off this time around. WDYT?

caseydavenport replied:
I think the issue with FelixConfiguration that makes my suggestion above not entirely viable is that the config that Felix ultimately sees might have both of these set, even if we prevent them from both being set on a single API object, since Felix merges config from multiple API objects as well as env vars, files, etc. So we need to think of this in two ways: What does the API do with this, and what does Felix do with this.

I think we should do effectively what Shaun suggested above, but with a little bit of API validation so that on a single object only one of these is ever set. i.e.,

- The API validation code should prevent both of these from being set on the same object (libcalico-go validation func) 
- Remove Felix config_param defaulting per @fasaxc's suggestion above. 
- Felix code should emit a warning log if it sees both config options as being set.
- Felix code should emit a warning log if it sees the old, deprecated option as being set.
- Felix code prefers the new option if both are set.

I think that means you probably don't need to re-work this logic again.

aaaaaaaalex replied:
Sounds reasonable! Will implement :smiley: thanks guys

fasaxc commented on calico/reference/resources/felixconfig.md:
```diff
@@ -82,7 +82,8 @@ spec:
 | reportingTTL                       | Time-to-live setting for process-wide status reports. | `5s`, `10s`, `1m` etc. | duration | `90s` |
 | routeRefreshInterval               | Period at which Felix re-checks the routes in the dataplane to ensure that no other process has accidentally broken {{site.prodname}}'s rules. Set to 0 to disable route refresh. | `5s`, `10s`, `1m` etc. | duration | `90s` |
-| routeTableRange                    | Calico programs additional Linux route tables for various purposes.  `RouteTableRange` specifies the indices of the route tables that Calico should use. |  | [RouteTableRange](#routetablerange) | `{Min: 1, Max: 250}` |
+| routeTableRange                    | *deprecated in favour of `RouteTableRanges`* Calico programs additional Linux route tables for various purposes. `RouteTableRange` specifies the indices of the route tables that Calico should use. If explicitly-set, will be honoured in favour of `RouteTableRanges`. |  | [RouteTableRanges](#routetablerange) | `{Min: 1, Max: 250}` |
```
Left the default on this one.

aaaaaaaalex replied:
good catch, thought I got them all :laughing:

caseydavenport replied:
Warning: American incoming

"honoured -> honored"

"favour" -> "favor" 

alternatively, we can avoid those contentious words altogether :laughing: 

> If explicitly set, this field will take precendence over RouteTableRanges

Non-american comment: "explicitly-set" shouldn't be hyphenated. "explicitly set" is fine.

lwr20 replied:
s/precendence/precedence

aaaaaaaalex replied:
cleaned up this wording now, although worth noting I saw a few unrelated parts of our codebase that still use european spelling, I dont know if a formal decision has been made around stuff like this (?), but maybe another, more intentional PR could grab those remaining bits if we want to make a concerted effort there

fasaxc reviewed (changes requested):
Basically looking good, couple of nits.

You should also add a release note (by editing the block in the top-level PR comment) since this deprecates a field.

aaaaaaaalex commented on libcalico-go/lib/validator/v3/validator.go:
```diff
@@ -772,6 +779,11 @@ func validateFelixConfigSpec(structLevel validator.StructLevel) {
+
+	if c.RouteTableRange != nil && c.RouteTableRanges != nil {
+		structLevel.ReportError(reflect.ValueOf(c.RouteTableRange),
+			"RouteTableRange", "", reason("cannot be set when `RouteTableRanges` is also set"), "")
```
@caseydavenport is this roughly what you had in mind when you mentioned the api faililng?
Plus corresponding test [here](https://github.com/projectcalico/calico/pull/5432/files#diff-d0fed76674ffad43c00300df4bc3af2c405bf6182041614bd9b5c62b678bc9ccR685)

fasaxc commented on felix/config/config_params.go:
```diff
@@ -855,6 +859,27 @@ func (config *Config) TyphaDiscoveryOpts() []discovery.Option {
+// RouteTableIndices compares provided args for the deprecated RoutTableRange arg
+// and the newer RouteTableRanges arg, giving precedence to the newer arg if it's explicitly-set
+func (config *Config) RouteTableIndices() []idalloc.IndexRange {
+	if config.RouteTableRanges == nil || len(config.RouteTableRanges) == 0 {
```
In Go, you don't need the separate nil check. `len(nil slice)` is 0

fasaxc commented on calico/reference/resources/felixconfig.md:
```diff
@@ -150,16 +150,15 @@ The `RouteTableRange` option is now deprecated in favour of [RouteTableRanges](#
 | max      | Maximum index to use | int    |
 
-Each item in the `RouteTableRanges` list designates a range of routing tables available to Calico. By default, Calico will use a single range of `1-10000`, and will automatically ignore Linux's default reserved tables `253-255`. However, it's possible that other table ranges may also be reserved by third-party systems unknown to Calico. In that case, multiple ranges can be defined to target table ...
+Each item in the `RouteTableRanges` list designates a range of routing tables available to Calico. By default, Calico will use a single range of `1-250`, and will automatically ignore Linux's default reserved tables `253-255`. However, it's possible that other table ranges may also be reserved by third-party systems unknown to Calico. In that case, multiple ranges can be defined to target tables  ...
```
```suggestion
Each item in the `RouteTableRanges` list designates a range of routing tables available to Calico. By default, Calico will use a single range of `1-250`.  If a range spans Linux's reserved table range (`253-255`) then those tables are automatically excluded from the list. It's possible that other table ranges may also be reserved by third-party systems unknown to Calico. In that case, multiple ranges can be defined to target tables below and above the sensitive ranges:
```

fasaxc commented on calico/reference/resources/felixconfig.md:
```diff
@@ -150,16 +150,15 @@ The `RouteTableRange` option is now deprecated in favour of [RouteTableRanges](#
 *Note*, for performance reasons, the maximum total number of routing tables that Felix will accept is 65535 (or 2*16).
 
-If the deprecated `RouteTableRange` option is explicitly set, it will be honoured in favour of `RouteTableRanges`.
+`RouteTableRanges` will override `RouteTableRange` if both are set.
```
```suggestion
If both `RouteTableRanges` and `RouteTableRange` are set, `RouteTableRanges` takes precedence and `RouteTableRange` is ignored.
```

fasaxc commented on calico/reference/resources/felixconfig.md:
```diff
@@ -150,16 +150,15 @@ The `RouteTableRange` option is now deprecated in favour of [RouteTableRanges](#
 | min      | Minimum index to use | int    |
```
Add the Accepted values column here?

fasaxc reviewed (approved):
Approved with a few small suggestions

aaaaaaaalex commented:
/merge-when-ready

aaaaaaaalex commented:
/merge-when-ready

aaaaaaaalex commented:
/merge-when-ready
